Support configuring "root element" for modal

spicalous opened the pull request:
Hi guys,

New contributor here, please let me know if I have missed anything

This PR attempts to resolve https://github.com/twbs/bootstrap/issues/31509 adding support to allow configuring the "root element" for the modal component. This PR only implements configuring this when initialising the modal via Javascript. I haven't dug deep into how this can be done via providing html attributes. 

Please let me know if this is the right approach


Thanks!

[preview](https://deploy-preview-33018--twbs-bootstrap.netlify.app/docs/5.0/components/modal/#configuring-root-element)


closes #34309

spicalous commented:
Hi @XhmikosR is there something I can do to help move this PR along? :)

XhmikosR commented:
No, sorry, I'm a little busy with Real Life.

Maybe @rohit2sharma95 or @GeoSot can help with review.

For what is worth, I'm definitely in favor of the null tests change.

GeoSot commented:
@spicalous try to create a working optical example in `site/content/docs/5.0/components/modal.md` in order to check it [here](https://deploy-preview-33018--twbs-bootstrap.netlify.app/docs/5.0/components/modal/)

_(You may need to revert it later)_

spicalous commented:
@XhmikosR I didn't mean to rush, thank you for your time and bringing relevant people to the discussion! 

@GeoSot I have added an example, please let me know what you think 
https://deploy-preview-33018--twbs-bootstrap.netlify.app/docs/5.0/components/modal/#configuring-root-element

GeoSot commented on site/assets/js/application.js:
```diff
@@ -101,6 +101,16 @@
+      rootElementModal.show()
+    })
+  }
+
```
I think, you can omit this and put a `data-bs-parent` attribute  (instead of `rootElement`)

GeoSot commented:
I would suggest you to open another Pr with the _null changes_ on tests 

I am not sure if `CLASS_NAME_OPEN_CUSTOM_ROOT`  is needed, at least yet. I understand the root element, but I don't know if its ok to restrict the modal in inside a box, But I am not the right guy to answer this

spicalous commented:
I have created https://github.com/twbs/bootstrap/pull/33288 separately.

In the mean time, i'm happy to either decline this or wait until someone can make the decision whether this feature should be added :)

Thanks for your time everyone

GeoSot commented:
> 
> 
> I have created #33288 separately for the tests changes

Spicalous Please remove them from here :)

GeoSot commented:
@spicalous I apologize for the delay. I had in mind this PR, and tried to make some steps, on backdrop, scrollbar and modal. Do you still want to close it?

spicalous commented:
Thanks @GeoSot, happy to re-open. Only just saw the mention from https://github.com/twbs/bootstrap/pull/33665 now. 

I haven't looked at the changes, but I assume we'll be able to incorporate the solution into this PR? I'll have a look at understanding the changes and will update this branch accordingly

Please let me know if this is what you had in mind or if there's something I should do / can help out with :)

GeoSot commented:
Scrollbar and backdrop functionality, have been decoupled to be used on offcanvas too,  
Backdrop is ready to accept a `rootElement`, but for scrollbar, I just needed to do it step by step. So I want to try to make the first step on  #33665. Please keep an eye on it.

GeoSot commented:
@spicalous 

I  think we are in the right moment to continue this MR.
Are ok to continue your proposal?

spicalous committed:
configurable modal custom root element

spicalous commented:
@GeoSot Had another attempt with the latest master

https://deploy-preview-33018--twbs-bootstrap.netlify.app/docs/5.0/components/modal/#configuring-root-element

Notes
- The css doesn't support configurable root elements out of the box... and css is not my strong suite - I think I will need some help to make it better as it feels like a hack at the moment or it will need some testing
- `.modal-open` seems to be an unused class now? So I have removed it

GeoSot commented:
>.modal-open seems to be an unused class now? So I have removed it

bring back .modal-open as we keep it for compatibility reasons

>The css doesn't support configurable root elements out of the box... and css is not my strong suite - I think I will need some help to make it better as it feels like a hack at the moment or it will need some testing


I 'll try to help you till there, and we will find a solution

Plus:

* Try to replace the `_isCustomRoot` with a private method `hasCustomRoot`
* try to resolve rootElement inside _getConfig (`getElement(config.rootElement)`)
* Scrollbar maybe will need something more, but not too much (leave it for the end 😉 )

spicalous committed:
PR amendment:
- resolve rootElement using getElement() when creating config;
- reinstate modal-open class
- refactor boolean flag to method

spicalous commented on js/tests/integration/bundle.js:
```diff
@@ -1,4 +1,4 @@
-import { Tooltip } from '../../../dist/js/bootstrap.esm.js'
+import { Tooltip } from './../../../dist/js/bootstrap.esm.js'
```
I had to add this, otherwise the build was failing for me :/

GeoSot commented:
Please, keep calm. I am here 😋 

Do not mess with backdrop yet. I am trying to help merging #34149

Try to keep your changes to minimum

GeoSot commented on js/src/modal.js:
```diff
@@ -35,13 +36,15 @@ const ESCAPE_KEY = 'Escape'
   keyboard: 'boolean',
-  focus: 'boolean'
+  focus: 'boolean',
+  rootElement: '(element|string)'
```
I think is enough as you resolve it before type check
```suggestion
  rootElement: 'element'
```

spicalous replied:
This is also the case in `backdrop.js` should I remove `string` from there also?

GeoSot replied:
I would suggest not to mess with backdrop at all for now 😉  (act like it is fine, in order to avoid open two many things at once)

spicalous replied:
https://github.com/twbs/bootstrap/pull/34149 is merged now, should I update this?

GeoSot commented on site/assets/js/application.js:
```diff
@@ -101,6 +101,16 @@
     })
   }
 
+  var rootElementExample = document.getElementById('modal-root-example')
```
you can avoid all this change using the `data-bs-root-element` markup on docs

spicalous replied:
Can you give an example? I can't see `data-bs-root-element` anywhere in code to see what it does

1. Is it something you suggest I implement?
2. Or is it implemented already like below?
```
<div id="modal-root-example" class="modal fade" data-bs-root-element="#another-element" ... >
```

GeoSot replied:
Number 2. 

```js 


  _getConfig(config) {
    config = {
      ...Default,
      ...Manipulator.getDataAttributes(this._element),
      ...(typeof config === 'object' ? config : {})
    }
    typeCheckConfig(NAME, config, DefaultType)
    return config
  }
  
 
```
Make a check, first but I am pretty sure, `Manipulator.getDataAttributes` resolves automatically element 
`data-bs-root-element` to  `this._config.rootElement`

spicalous replied:
Yep, that looks good! However `Manipulator.getDataAttributes()` will return `rootElement` as a string. I have kept the `getElement` method call to resolve the string selector to an element

GeoSot commented:
@spicalous any update on this?

spicalous committed:
remove string from type check as it is already resolved

spicalous committed:
revert: modal-open comment removal

GeoSot commented on js/src/modal.js:
```diff
@@ -240,6 +247,9 @@ class Modal extends BaseComponent {
 
     this._element.classList.add(CLASS_NAME_SHOW)
+    if (this._hasCustomRoot()) {
+      this._element.classList.add(CLASS_NAME_MODAL_CUSTOM_ROOT)
```
Maybe you can avoid these custom root classes and checks, by simple using a css selector like `.modal-open:not(body)`

spicalous replied:
Good idea, amended.

GeoSot commented on js/src/util/backdrop.js:
```diff
@@ -74,7 +75,9 @@ class Backdrop {
     if (!this._element) {
       const backdrop = document.createElement('div')
-      backdrop.className = this._config.className
+      backdrop.className = this._hasCustomRoot() ?
```
Same as modal custom root class

spicalous committed:
use data-bs-root-element; replace custom classes with .modal-open:not(body)

GeoSot commented on js/src/util/scrollbar.js:
```diff
@@ -13,8 +13,8 @@ const SELECTOR_FIXED_CONTENT = '.fixed-top, .fixed-bottom, .is-fixed, .sticky-to
-  constructor() {
-    this._element = document.body
+  constructor(rootElement) {
+    this._element = rootElement || document.body
```
You miss some things here :wink: 

plus the proper tests

This [MR](https://github.com/twbs/bootstrap/pull/33665/files) will help you

spicalous replied:
I'm not sure how I could add a test in modal (I see in the MR that we are asserting spied methods of `ScrollBarHelper`. Please let me know if you had something in mind. In the mean time, I have added a test for `ScrollBarHelper`

GeoSot replied:
I tried to fix it a bit and added the tests and functionality for scrollbar. 
I am not very sure about scrollbar modifications, so you better take a second look

GeoSot commented:
Nice job till here :+1:  Getting closer

modal.js:l89 misses the proper initialization

spicalous committed:
amended scroll bar instantiation with missed root element

GeoSot committed:
Add scrollbar functionality for different rootElements than body

GeoSot commented:
I found out a 'bug' 

if you try to put some content inside the empty root element,  it will over-float. Till here we are fine. 
Although if you try open the modal inside the scrolled down root element, the absolute position, will be out of user eyes :confused:

GeoSot commented:
I am closing this as stale. 
@spicalous if you find time, you can reopen it
